consensus/babe/babe.go

```diff
@@ -40,3 +50,67 @@ func NewBabeSession(pubkey VrfPublicKey, privkey VrfPrivateKey, rt *runtime.Runt
 		rt:            rt,
 	}
 }
+
+// runs the slot lottery for a specific slot
+// returns true if validator is authorized to produce a block for that slot, false otherwise
+func (b *BabeSession) runLottery(slot uint64) (bool, error) {
+	if slot < b.epochData.StartSlot {
+		return false, errors.New("slot is not in this epoch")
+	}
+
+	output, err := b.vrfSign(slot)
+	if err != nil {
+		return false, err
+	}
+
+	output_int := new(big.Int).SetBytes(output)
+	threshold, err := calculateThreshold(b.config.C1, b.config.C2, b.authorityIndex, b.authorityWeights)
```

Comment:
Perhaps this should be calculated and stored?

Reply:
good move, I've changed the code to calculate it once for the epoch and store in the BabeSession